File: solutions/0238. Product of Array Except Self/0238.cpp
// Time complexity: O(n)
// Space complexity: O(n)
class Solution {
public:
    vector<int> productExceptSelf(const vector<int>& nums) {
        const int n = nums.size();
        vector<int> ans(n, 1);

        for (int i = 1; i < n; ++i) {
            ans[i] = ans[i - 1] * nums[i - 1];
        }

        int suffix = 1;
        for (int i = n - 2; i >= 0; --i) {
            suffix *= nums[i + 1];
            ans[i] *= suffix;
        }

        return ans;
    }
};

File: solutions/0238. Product of Array Except Self/0238.rs
// Time complexity: O(n)
// Space complexity: O(n)
impl Solution {
    pub fn product_except_self(nums: Vec<i32>) -> Vec<i32> {
        let n = nums.len();
        let mut ans = vec![1; n];

        for i in 1..n {
            ans[i] = ans[i - 1] * nums[i - 1];
        }

        let mut suffix = 1;
        for i in (0..n - 1).rev() {
            suffix *= nums[i + 1];
            ans[i] *= suffix;
        }

        ans
    }
}